packages/chatbot-server-mongodb-public/src/eval/ConversationEval.ts

```diff
@@ -185,6 +199,246 @@ const makeFactuality: ConversationEvalScorerConstructor =
+      responseScores?.assessments;
+    assert(assessments.length > 0, "No assessments found");
+
+    return {
+      name,
+      score:
+        assessments.reduce(
+          (accumulator, current) => accumulator + current.score,
+          0
+        ) / args.expected?.promptAdherence.length,
+      metadata: {
+        scores: response.choices[0].message.content,
+      },
+    };
+  };
+};
+
+const ToolCallAmountCorrect: ConversationEvalScorer = (args) => {
+  const name = "ToolCallAmountCorrect";
+  if (args.expected?.messages === undefined) {
+    return {
+      name,
+      score: null,
+    };
+  }
+
+  const expectedToolCallCount = args.expected.messages.filter(
+    (message) => message.role === "assistant-tool"
+  ).length;
+  const actualToolCallCount = args.output.messages.filter(
+    (message) => message.role === "assistant" && message?.toolCall !== undefined
+  ).length;
+
+  if (expectedToolCallCount === 0) {
+    return {
+      name,
+      score: actualToolCallCount === 0 ? 1 : 0,
+      metadata: {
+        actualToolCallCount,
+        expectedToolCallCount,
+      },
+    };
+  }
+  return {
+    name,
+    score: actualToolCallCount / expectedToolCallCount,
+    metadata: {
+      actualToolCallCount,
+      expectedToolCallCount,
+    },
+  };
+};
+
+/** Verify the messages called in the correct order with the correct tool calls */
+const MessageOrderCorrect: ConversationEvalScorer = (args) => {
+  const name = "MessageOrderCorrect";
+  if (args.expected?.messages === undefined) {
+    return {
+      name,
+      score: null,
+    };
+  }
+
+  if (args.output.messages.length !== args.expected.messages.length) {
+    return {
+      name: name,
+      score: 0,
+      metadata: {
+        message: "Output and expected messages length were different",
+      },
+    };
+  }
+
+  const score =
+    args.output.messages
+      .map((outputMessage, idx) => {
+        const expectedMessage = args.expected.messages?.[idx];
+
+        // If the expected message is undefined, score 0
+        if (!expectedMessage) {
+          return 0;
+        }
+
+        const sameRole =
+          (expectedMessage.role === "assistant-tool" &&
+            outputMessage.role === "assistant") ||
+          expectedMessage.role === outputMessage.role;
+
+        // If different role, score 0
+        if (!sameRole) {
+          return 0;
+        }
+
+        // Also validate the correct tool was called (by name)
+        const hasExpectedToolCall =
+          outputMessage.role === "assistant"
+            ? outputMessage.toolCall?.function.name ===
+              expectedMessage.toolCallName
+            : true;
+
+        const messageScore: number = sameRole && hasExpectedToolCall ? 1 : 0;
+        return messageScore;
+      })
+      .reduce((acc, score) => acc + score, 0) / args.output.messages.length;
+
+  return {
+    name: name,
+    score,
+    metadata: {
+      message: `Scored ${score} based on matching roles`,
+    },
```

Comment:
this metadata seems low value. maybe instead could return the total num matches and the total num messages:

```ts
{
  //...
  metadata: { numMatches: 3, totalMessages: 5}
}
```